merkle/maphasher/generic.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package maphasher provides hashing for maps.
+package maphasher
+
+import (
+	"crypto"
+	"fmt"
+)
+
+// Domain separation prefixes
+const (
+	leafHashPrefix = 0
+	nodeHashPrefix = 1
+)
+
+// Default is a SHA256 based TreeHasher for maps.
+var Default = New(crypto.SHA256)
+
+// TreeHasher implements the RFC6962 tree hashing algorithm.
+// Empty branches within the tree are plain interior nodes e1 = H(e0, e0) etc.
+type TreeHasher struct {
+	crypto.Hash
+	nullHashes [][]byte
+}
+
+// New creates a new TreeHasher on the passed in hash function.
+func New(h crypto.Hash) *TreeHasher {
+	m := &TreeHasher{Hash: h}
+	m.nullHashes = m.createNullHashes()
+	return m
+}
+
+// String returns a string representation for debugging.
+func (t *TreeHasher) String() string {
+	return fmt.Sprintf("genericHash{%v}", t.Hash)
+}
+
+// EmptyRoot returns the root of an empty tree.
+func (t *TreeHasher) EmptyRoot() []byte {
+	return t.HashEmpty(t.Size() * 8)
+}
+
+// HashEmpty returns the hash of an empty branch at a given depth.
+// A depth of 0 indictes the hash of an empty leaf.
+func (t *TreeHasher) HashEmpty(depth int) []byte {
+	if depth < 0 || depth >= len(t.nullHashes) {
+		panic(fmt.Sprintf("HashEmpty(%v) out of bounds", depth))
```

Comment:
This is more of an assert to make the out of bounds error in the next statement more readable.  We should only encounter this error if there's a logic bug elsewhere in the code. I'm happy to leave it as a panic to avoid the full error test branching throughout the code. What do you think?